feat(draw): add basic vector APIs

PeterBee97 opened the pull request:
### Description of the feature or fix
Vector APIs and basic draw task distribution following #4388 . Some matrix helper functions are provided.
### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

kisvegabor commented on src/draw/lv_draw_vector.c:
```diff
@@ -0,0 +1,144 @@
+void lv_vector_rotate(lv_draw_vector_dsc_t * dsc, float degree, lv_coord_t pivot_x, lv_coord_t pivot_y)
+{
+    /* right multiply by [c -s (1-c)x+sy; s c (1-s)x-cy; 0 0 1] */
+    float radian = degree / 180.0f * M_PI;
```
```suggestion
    float radian = degree / 180.0f * (float)M_PI;
```

PeterBee97 replied:
fixed

kisvegabor commented:
Hey,

Thank you! 

Have you already tried it out with ThorVG or VG-Lite (or other GPU)?

XuNeo commented on src/draw/lv_draw_vector.c:
```diff
@@ -0,0 +1,144 @@
+    LV_PROFILER_END;
+}
+
+void lv_draw_vector(struct _lv_layer_t * layer, lv_vector_path_t * path, lv_draw_vector_dsc_t * dsc)
```
```suggestion
void lv_draw_vector(struct _lv_layer_t * layer, lv_draw_vector_dsc_t * dsc, lv_vector_path_t * path)
```

`lvgl` follow the pattern of draw(layer, dsc, other_args).
See:
```c

/**
 * Draw an image
 * @param draw_ctx      pointer to the current draw context
 * @param dsc           pointer to an initialized `lv_draw_img_dsc_t` variable
 * @param coords        the coordinates of the image
 * @param src           pointer to a lv_color_t array which contains the pixels of the image
 */
void lv_draw_img(struct _lv_layer_t * layer, const lv_draw_img_dsc_t * dsc, const lv_area_t * coords);


void lv_draw_layer(struct _lv_layer_t * layer, const lv_draw_img_dsc_t * dsc, const lv_area_t * coords);
LV_ATTRIBUTE_FAST_MEM void lv_draw_label(lv_layer_t * layer, const lv_draw_label_dsc_t * dsc,
                                         const lv_area_t * coords);

LV_ATTRIBUTE_FAST_MEM void lv_draw_letter(lv_layer_t * layer, lv_draw_label_dsc_t * dsc,
                                          const lv_point_t * point, uint32_t unicode_letter);

```

PeterBee97 replied:
done

XuNeo commented on src/draw/lv_draw_vector.c:
```diff
@@ -0,0 +1,144 @@
+ *  STATIC FUNCTIONS
+ **********************/
+
+static inline void _matrix_translate(float m[3][3], float x, float y)
```
Let's move to header files in case widgets may need it?

onecoolx replied:
> Let's move to header files in case widgets may need it?

I think we should add a set of matrix api like  
lv_matrix_translate
lv_matrix_scale
.....
Because some GPUs can provide the interface of matrix transformation 
For example:
               vg_lite_rotate
               vg_lite_scale
               ....
Or we can also acceleration at the bottom with  NEON Instrinsic.

PeterBee97 replied:
I agree with bringing the declarations to public, however since vg_lite matrix APIs are naive matrix multiplications I would advice against using them.

XuNeo commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,145 @@
+};
+typedef uint8_t lv_vector_stroke_join_style_t;
+
+typedef struct {
+    float m[3][3];
+} lv_matrix_t;
```
```suggestion
typedef float (lv_matrix_t)[3][3];
```

PeterBee97 replied:
As per https://stackoverflow.com/a/4523537 I would like to keep current implementations, what do you think?

kisvegabor commented:
As these files uses `float` please add an `#if LV_USE_VECTOR` to let the user make LVGL work without `float`.

PeterBee97 commented on src/draw/lv_draw_vector.c:
```diff
@@ -0,0 +1,144 @@
+{
+    LV_PROFILER_BEGIN;
+    lv_area_t a;
+    if(!_lv_area_intersect(&a, &layer->clip_area, &dsc->scissor_area)) return;
```
moved LV_PROFILER_BEGIN to after the return

PeterBee97 commented on src/draw/lv_draw.h:
```diff
@@ -48,6 +48,8 @@ typedef enum {
     LV_DRAW_TASK_TYPE_TRIANGLE,
     LV_DRAW_TASK_TYPE_MASK_RECTANGLE,
     LV_DRAW_TASK_TYPE_MASK_BITMAP,
+    LV_DRAW_TASK_TYPE_VECTOR_FILL,
```
sure

PeterBee97 commented:
> Hey,
> 
> Thank you!
> 
> Have you already tried it out with ThorVG or VG-Lite (or other GPU)?

Not yet, this is only the task distribution part and no task handler has been implemented so far.

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,188 @@
+/*Helper functions*/
+void lv_matrix_translate(lv_matrix_t * matrix, float dx, float dy);
+void lv_matrix_scale(lv_matrix_t * matrix, float scale_x, float scale_y);
+void lv_matrix_rotate(lv_matrix_t * matrix, float cosine, float sine);
```
void lv_matrix_rotate(lv_matrix_t * matrix, float degree);

PeterBee97 commented:
Updated:
1. Changed all API names to lv_vpath_*
2. Added lv_vpath_submit to daisy chain draw tasks before submitting them. GPU will benefit from batch processing.
3. Enriched matrix helpers with pseudo 3d transformations

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+void lv_vpath_set_order(lv_vpath_dsc_t * dsc, bool stroke_first);
+
+/*Draw path commands*/
+lv_vpath_task_t * lv_vpath_task_create(void);
```
**lv_vpath_task_t**  needs to be an private data and should not be created and maintained by the API caller

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+void lv_vpath_draw(struct _lv_layer_t * layer, lv_vpath_task_t * task, lv_vpath_t * path, lv_vpath_dsc_t * dsc);
+void lv_vpath_clear(struct _lv_layer_t * layer, lv_vpath_task_t * task, lv_area_t * rect, lv_color_t color);
+void lv_vpath_submit(struct _lv_layer_t * layer, lv_vpath_task_t * task);
+void lv_vpath_draw_oneshot(struct _lv_layer_t * layer, lv_vpath_t * path, lv_vpath_dsc_t * dsc);
```
why need this？

onecoolx replied:
“void lv_vpath_draw_oneshot(struct _lv_layer_t * layer, lv_vpath_t * path, lv_vpath_dsc_t * dsc);”
why need this？

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+void lv_matrix_scale(lv_matrix_t * matrix, float scale_x, float scale_y);
+void lv_matrix_rotate(lv_matrix_t * matrix, float angle);
+/*Pseudo 3D rotation (using affine transformations to simulate 3d rotations cast to xy plane)*/
+void lv_matrix_rotate_p3d(lv_matrix_t * matrix, float rx, float ry, float rz);
```
“
void lv_matrix_rotate_p3d(lv_matrix_t * matrix, float rx, float ry, float rz);
void lv_matrix_rotate_quaternion(lv_matrix_t * matrix, float q0, float q1, float q2, float q3);
”
If you want to implement 3D transformation, you need **matrix[4][4]**, not **matrix[3][3]**

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+/*Helper functions*/
+void lv_matrix_translate(lv_matrix_t * matrix, float dx, float dy);
+void lv_matrix_scale(lv_matrix_t * matrix, float scale_x, float scale_y);
+void lv_matrix_rotate(lv_matrix_t * matrix, float angle);
```
Missing matrix skew method

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+void lv_vpath_append_path(lv_vpath_t * path, lv_vpath_t * subpath);
+
+/*Param operations*/
+void lv_vpath_dsc_init(lv_vpath_dsc_t * dsc);
```
```suggestion
void lv_vpath_dsc_init(lv_vpath_dsc_t * dsc);
```
In order to save the **lv_vpath_task_t** we can change **lv_vpath_dsc_t** to **lv_vpath_draw_context**
That we can create and maintain it using methods like:
```
lv_vpath_draw_context*  lv_vpath_draw_context_create(struct _lv_layer_t * layer);
lv_vpath_draw_context_destroy(lv_vpath_draw_context* ctx)
```
The drawing function can be:
```
void lv_vpath_draw( lv_vpath_draw_context * ctx, lv_vpath_t * path);
```

onecoolx replied:
@PeterBee97  Do you think this is better?

onecoolx commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+    lv_vpath_blend_t blend;
+    lv_area_t scissor_area;
+    bool stroke_first;
+} lv_vpath_dsc_t;
```
```suggestion
typedef struct {
    lv_vpath_fill_dsc_t fill_dsc;
    lv_vpath_stroke_dsc_t stroke_dsc;
    lv_matrix_t matrix;
    lv_vpath_blend_t blend;
    lv_area_t scissor_area;
    bool stroke_first;
} lv_vpath_dsc_t;
```
```
typedef struct {
    lv_vpath_fill_dsc_t fill_dsc;
    lv_vpath_stroke_dsc_t stroke_dsc;
    lv_matrix_t matrix;
    lv_vpath_blend_t blend;
    lv_area_t scissor_area;
    bool stroke_first;
} lv_vpath_dsc_t;

typedef struct {
    lv_vpath_dsc_t draw_dsc;
    lv_vpath_task_t * task;
    lv_vpath_task_t * current_task;
} lv_vpath_draw_context_t;
```

kisvegabor commented:
Please use `vector_path` instead of `vpath`. From v9 abbreviations should be avoided for the sake of clarity.

onecoolx commented:
> Please use `vector_path` instead of `vpath`. From v9 abbreviations should be avoided for the sake of clarity.
I think "lv_vector_path_"  is too long as a prefix, how about “lvg_" ？  @kisvegabor

kisvegabor commented:
I looked into the API again  I'm a little bit puzzled by
- `lv_vpath_draw`
- `lv_vpath_clear`
- `lv_vpath_submit`
- `lv_vpath_draw_oneshot`
- `lv_vpath_task_create`

have the same prefix as `lv_vpath_quad_to`, but they operate on a task list and not a vector path. 

What about something like this:
```c
/*Same as `lv_vpath_t`
 *"vector_path" is not prefix but the whole name here*/
lv_vector_path_t * lv_vector_path_create(void);
void lv_vector_path_quad_to(lv_vector_path_t * path, lv_fpoint_t p1, lv_fpoint_t p2);

/*Same as `lv_vpath_dsc_t`
 *lv_vector_draw_dsc_t ) has   `lv_vector_path_t * path`  and  `lv_vector_draw_dsc_t * next` fields*/
void lv_vector_draw_dsc_init(lv_vector_draw_dsc_t * dsc); 
dsc.path = some_path;
dsc.fill_dsc.color = lv_color_hex(0xffaa00);
dsc.next = &another_dsc;

/*Start drawing*/
void lv_vector_draw(struct _lv_layer_t * layer, lv_vector_draw_dsc_t * dsc);
```

onecoolx commented:
> I looked into the API again I'm a little bit puzzled by
> 
> * `lv_vpath_draw`
> * `lv_vpath_clear`
> * `lv_vpath_submit`
> * `lv_vpath_draw_oneshot`
> * `lv_vpath_task_create`
> 
> have the same prefix as `lv_vpath_quad_to`, but they operate on a task list and not a vector path.
> 
> What about something like this:
> 
> ```c
> /*Same as `lv_vpath_t`
>  *"vector_path" is not prefix but the whole name here*/
> lv_vector_path_t * lv_vector_path_create(void);
> void lv_vector_path_quad_to(lv_vector_path_t * path, lv_fpoint_t p1, lv_fpoint_t p2);
> 
> /*Same as `lv_vpath_dsc_t`
>  *lv_vector_draw_dsc_t ) has   `lv_vector_path_t * path`  and  `lv_vector_draw_dsc_t * next` fields*/
> void lv_vector_draw_dsc_init(lv_vector_draw_dsc_t * dsc); 
> dsc.path = some_path;
> dsc.fill_dsc.color = lv_color_hex(0xffaa00);
> dsc.next = &another_dsc;
> 
> /*Start drawing*/
> void lv_vector_draw(struct _lv_layer_t * layer, lv_vector_draw_dsc_t * dsc);
> ```


I think this way is not easy to use. The draw_dec list should be implements under a contxt and should not be maintained by the caller, which is easy cause to errors. 
for example:
```
lv_vpath_draw_context_t* ctx = lv_vpath_draw_context_create(layer);

/*Start drawing*/
lv_vector_path_set_fill_color(ctx, lv_color_hex(0xffaa00));
lv_vector_draw(ctx,  some_path);

lv_vector_path_set_stroke_color(ctx, lv_color_hex(0xffbbaa));
lv_vector_draw(ctx,  another_path);

lv_vector_commit(ctx); // submit the draw list here

...
lv_vpath_draw_context_destroy(ctx);
```

In this way, the caller of the API does not need to worry about drawing the list and can focus on the drawing logic. 
What do you think of this way? @kisvegabor

kisvegabor commented:
What is the content of `lv_vpath_draw_context_t`? Is it different from `lv_vpath_dsc_t` in the PR?

So `lv_vector_draw` handles the creation of the list, right?


For the sake of consistency with other draw tasks, it'd be great if `lv_vector_draw` could mean the real drawing (`lv_vector_commit` in the example) and `lv_vector_commit` would be `lv_vector_add` or something like that.

onecoolx commented:
> What is the content of `lv_vpath_draw_context_t`? Is it different from `lv_vpath_dsc_t` in the PR?
> 
> So `lv_vector_draw` handles the creation of the list, right?
> 
> For the sake of consistency with other draw tasks, it'd be great if `lv_vector_draw` could mean the real drawing (`lv_vector_commit` in the example) and `lv_vector_commit` would be `lv_vector_add` or something like that.

`lv_vpath_draw_context_t` is used to save and maintain 'draw_list'.  When `lv_vector_draw` is called, a new draw operation will be added to 'draw_list' , When `lv_vector_commit` is called, a real lvgl  draw task will be created for the drawing task, and the 'draw_list' will be traversed in callback to complete the actual drawing. It is similar to a "Recording / Playback" mechanism. 
We think this would reduce interruptions to other lvgl widget's drawing.

PeterBee97 commented:
Refined API naming and added skew functions. There is one more thing from our internal discussion:
Since the parallel rendering architecture is asynchronous, the path content has to be copied when drawing. As such, no zero-copying of path can be achieved, it shouldn't be much pain to add an internal representation of path and convert it to vendor-specific format during rendering. (It's O(n) anyway)

kisvegabor commented on src/draw/lv_draw_vector.h:
```diff
@@ -0,0 +1,213 @@
+void lv_vector_path_append_path(lv_vector_path_t * path, lv_vector_path_t * subpath);
+
+/*Context operations*/
+lv_vector_context_t * lv_vector_context_create(void);
```
E.g. in `lv_draw_img` we call it `lv_draw_image_dsc`. So wonder if we can use the `lv_vector_dsc_t` name, and make all related functions start with `lv_vector_dsc_*`. E.g. 
```c
 lv_vector_dsc_set_fill_color(dsc_t * dsc, lv_color_t color)
```

kisvegabor commented:
You mean `lv_vector_path_t`, right? If so, I think we can make the vector library to populate a `lv_vector_path_t` type. IMO it would be quite consistent as functions like `lv_vector_path_move_to` are implemented by the vector library anyway as well.

kisvegabor commented:
Not stale

kisvegabor commented:
Closing in favor of #4691.
